### DISCUSSION TITLE
A potential solution for AudioCache problems with Android API levels below 21

### DISCUSSION THREAD

**Discussion by sgg-labs** [bluefireteam/audioplayers#854] [General]
As a follow-up to #822 and a response to #850

I tried this modification in an API 19 emulator, and it seemed to work for playing short sound effect mp3 files. Unfortunately, I don't have a device to test on.

In WrappedSoundPool.kt, replace this
```
        private fun createSoundPool(): SoundPool {
            val attrs = AudioAttributes.Builder().setLegacyStreamType(AudioManager.USE_DEFAULT_STREAM_TYPE)
                    .setUsage(AudioAttributes.USAGE_GAME)
                    .build()
            return SoundPool.Builder()
                    .setAudioAttributes(attrs)
                    .setMaxStreams(100)
                    .build()
        }
```
with this
```
        private fun createSoundPool(): SoundPool {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) { 
                val attrs = AudioAttributes.Builder().setLegacyStreamType(AudioManager.USE_DEFAULT_STREAM_TYPE)
                        .setUsage(AudioAttributes.USAGE_GAME)
                        .build()
                return SoundPool.Builder()
                        .setAudioAttributes(attrs)
                        .setMaxStreams(100)
                        .build()
            } else {
                return SoundPool(100, AudioManager.STREAM_MUSIC, 0)
            }
        }
```

**Comment by baraaaljabban** [comment]
I tried your solution and it didn't work for me on test device ( Samsung one)

**Comment by sgg-labs** [comment]
**AudioAttributes** class was introduced in Android API level 21. I made its usage conditional in **WrappedSoundPool.kt**, but it's also used in **WrappedMediaPlayer.kt** (in **play** method and in **setAttrbutes** method). My test was limited to **AudioCache** in LOW_LATENCY mode, so it looks like the small change above was enough (at least on the emulator).

**Comment by erickzanardo** [comment]
This is already fixed by https://github.com/luanpotter/audioplayers/pull/841

This will be shipped on the next release.

**Reply by sgg-labs** [reply]
But #841 didn't change anything in the **WrappedSoundPool.kt**, which is what's used by **AudioCache**, if I understand it correctly. What am I missing here?

**Reply by erickzanardo** [reply]
AudioPool is only used when you use the low latency mode.

Which in turn cannot be supported on older devices because sound pool was introduced on API 23 if I remember correctly. So there isn't much to do regarding that.

But that PR makes audio cache works for the normal mode.

**Reply by sgg-labs** [reply]
And my little change above could make **AudioCache** work in low latency mode as well, because **SoundPool** dates all the way back to API 1, it's the **AudioAttributes** that didn't exist prior to API 21. I just don't have a physical device to test that this change doesn't lead to some funny things down the road. It worked ok in the API 19 emulator.

**Reply by erickzanardo** [reply]
Cool, would you be willing to make us a PR?

**Reply by sgg-labs** [reply]
If I do, it will be my first ever :)  It's a small change, maybe you take it from here?

**Reply by erickzanardo** [reply]
I will not have time to do that right now as I have other stuff in the queue.

Don't worry about it be your first PR, we can help guiding you if you have any doubts.

Simple changes makes the best first time PR :D
